metrics/local.go

```diff
@@ -0,0 +1,279 @@
+package metrics
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/codahale/hdrhistogram"
+)
+
+const (
+	defaultTick = time.Minute
+)
+
+// This is intentionally very similar to github.com/codahale/metrics, the
+// main difference being that counters/gauges are scoped to the provider
+// rather than being global (to facilitate testing).
+
+// A LocalBackend is a metrics provider which aggregates data in-vm, and
+// allows exporting snapshots to shove the data into a remote collector
+type LocalBackend struct {
+	cm       sync.RWMutex
+	gm       sync.RWMutex
+	tm       sync.RWMutex
+	counters map[string]*int64
+	gauges   map[string]*int64
+	timers   map[string]*localBackendTimer
+	stop     chan struct{}
+	stopped  chan struct{}
+}
+
+// NewLocalBackend returns a new LocalBackend
+func NewLocalBackend(tick time.Duration) *LocalBackend {
+	b := &LocalBackend{
+		counters: make(map[string]*int64),
+		gauges:   make(map[string]*int64),
+		timers:   make(map[string]*localBackendTimer),
+		stop:     make(chan struct{}),
+		stopped:  make(chan struct{}),
+	}
+
+	go func() {
+		if tick == 0 {
+			tick = defaultTick
+		}
+		ticker := time.NewTicker(tick)
+		for {
+			select {
+			case <-ticker.C:
+				b.tm.Lock()
+				timers := make(map[string]*localBackendTimer, len(b.timers))
+				for timerName, timer := range b.timers {
+					timers[timerName] = timer
+				}
+				b.tm.Unlock()
+
+				for _, t := range b.timers {
+					t.Lock()
+					t.hist.Rotate()
+					t.Unlock()
+				}
+			case <-b.stop:
+				ticker.Stop()
+				close(b.stopped)
+				return
+			}
+		}
+	}()
+	return b
+}
+
+// IncCounter increments a counter value
+func (b *LocalBackend) IncCounter(name string, tags map[string]string, delta int64) {
+	b.cm.RLock()
+	counter := b.counters[name]
+	b.cm.RUnlock()
+
+	if counter != nil {
+		atomic.AddInt64(counter, delta)
+		return
+	}
+
+	b.cm.Lock()
+	counter = b.counters[name]
+	if counter == nil {
+		b.counters[name] = new(int64)
+		*b.counters[name] = delta
+	} else {
+		atomic.AddInt64(counter, delta)
+	}
+	b.cm.Unlock()
+}
+
+// UpdateGauge updates the value of a gauge
+func (b *LocalBackend) UpdateGauge(name string, tags map[string]string, value int64) {
+	b.gm.RLock()
+	gauge := b.gauges[name]
+	b.gm.RUnlock()
+
+	if gauge != nil {
+		atomic.StoreInt64(gauge, value)
+		return
+	}
+
+	b.gm.Lock()
+	gauge = b.gauges[name]
+	if gauge == nil {
+		b.gauges[name] = new(int64)
+		*b.gauges[name] = value
+	} else {
+		atomic.StoreInt64(gauge, value)
+	}
+	b.gm.Unlock()
+}
+
+// RecordTimer records a timing duration
+func (b *LocalBackend) RecordTimer(name string, tags map[string]string, d time.Duration) {
+	timer := b.findOrCreateTimer(name)
```

Comment:
The tags are ignored here. That will make it impossible to use this local backend in unit tests that are sensitive to the tags. We can address this in a different PR, basically by copying functionality from the metrics implementation in jaeger-client-go.